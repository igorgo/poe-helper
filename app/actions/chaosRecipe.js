// @flow
import type { ChaosRecipeAction, ChaosRecipeItemType } from '../reducers/types';

export const INC_CHAOS_RECIPE = 'INC_CHAOS_RECIPE';
export const DEC_CHAOS_RECIPE = 'DEC_CHAOS_RECIPE';
export const RESET_CHAOS_RECIPE = 'RESET_CHAOS_RECIPE';
export const MINUS_SET_CHAOS_RECIPE = 'MINUS_SET_CHAOS_RECIPE';

export function increment(itemType: ChaosRecipeItemType): ChaosRecipeAction {
  return {
    type: INC_CHAOS_RECIPE,
    payload: {
      itemType
    }
  };
}

export function decrement(itemType: ChaosRecipeItemType): ChaosRecipeAction {
  return {
    type: DEC_CHAOS_RECIPE,
    payload: {
      itemType
    }
  };
}

export function reset(): ChaosRecipeAction {
  return {
    type: RESET_CHAOS_RECIPE
  };
}

export function minusSet(): ChaosRecipeAction {
  return {
    type: MINUS_SET_CHAOS_RECIPE
  };
}
