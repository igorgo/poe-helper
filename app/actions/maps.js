// @flow
import type { MapAction, MapId } from '../reducers/types';

export const MAP_ADD = 'MAP_ADD';
export const MAP_REMOVE = 'MAP_REMOVE';
export const MAP_SET_OWN = 'MAP_SET_OWN';
export const MAP_SET_SHAPE = 'MAP_SET_SHAPE';
export const MAP_SET_ELDER = 'MAP_SET_ELDER';
export const MAP_CLEAR = 'MAP_CLEAR';

export function mapAdd(id: MapId): MapAction {
  return {
    type: MAP_ADD,
    payload: {
      id
    }
  };
}

export function mapRemove(id: MapId): MapAction {
  return {
    type: MAP_REMOVE,
    payload: {
      id
    }
  };
}

export function mapSetOwn(id: MapId): MapAction {
  return {
    type: MAP_SET_OWN,
    payload: {
      id,
      own: true
    }
  };
}

export function mapUnsetOwn(id: MapId): MapAction {
  return {
    type: MAP_SET_OWN,
    payload: {
      id,
      own: false
    }
  };
}

export function mapSetShape(id: MapId): MapAction {
  return {
    type: MAP_SET_SHAPE,
    payload: {
      id,
      shape: true
    }
  };
}

export function mapSetElder(id: MapId): MapAction {
  return {
    type: MAP_SET_ELDER,
    payload: {
      id,
      elder: true
    }
  };
}

export function mapUnsetElder(id: MapId): MapAction {
  return {
    type: MAP_SET_ELDER,
    payload: {
      id,
      elder: false
    }
  };
}

export function mapUnsetShape(id: MapId): MapAction {
  return {
    type: MAP_SET_SHAPE,
    payload: {
      id,
      shape: false
    }
  };
}

export function clear(): MapAction {
  return {
    type: MAP_CLEAR
  };
}
