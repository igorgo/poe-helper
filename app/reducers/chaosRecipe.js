// @flow
import { remote } from 'electron';
import {
  INC_CHAOS_RECIPE,
  DEC_CHAOS_RECIPE,
  RESET_CHAOS_RECIPE,
  MINUS_SET_CHAOS_RECIPE
} from '../actions/chaosRecipe';
import type { ChaosRecipeAction, ChaosRecipeStore } from './types';
import { rows } from '../components/ChaosRecipe/ChaosRecipe';

const defaultState: ChaosRecipeStore = {
  weapon: 0,
  helmet: 0,
  body: 0,
  glove: 0,
  boot: 0,
  ring: 0,
  amulet: 0,
  belt: 0
};

const initialState: ChaosRecipeStore =
  (remote && remote.getGlobal('pstore').get('ChaosRecipe')) || defaultState;

const minusSet = (state: ChaosRecipeStore): ChaosRecipeStore => {
  const newState = {};
  Object.keys(state).forEach(i => {
    const newCount = state[i] - rows[i].length / 6;
    newState[i] = newCount < 0 ? 0 : newCount;
  });
  return newState;
};

export default function chaosRecipe(
  state: ChaosRecipeStore = initialState,
  action: ChaosRecipeAction
) {
  const itemType = action.payload ? action.payload.itemType : undefined;
  let newState = state;
  switch (action.type) {
    case INC_CHAOS_RECIPE:
      if (itemType && state[itemType] !== rows[itemType].length) {
        newState = { ...state, [itemType]: state[itemType] + 1 };
      }
      break;
    case DEC_CHAOS_RECIPE:
      if (itemType && state[itemType] !== 0) {
        newState = { ...state, [itemType]: state[itemType] - 1 };
      }
      break;
    case RESET_CHAOS_RECIPE:
      newState = defaultState;
      break;
    case MINUS_SET_CHAOS_RECIPE:
      newState = minusSet(state);
      break;
    default:
      newState = state;
  }
  if (remote) {
    remote.getGlobal('pstore').set('ChaosRecipe', newState);
  }
  return newState;
}
