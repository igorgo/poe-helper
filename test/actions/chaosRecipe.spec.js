import * as actions from '../../app/actions/chaosRecipe';

describe('actions', () => {
  it('should increment should create increment action', () => {
    expect(actions.increment('body')).toMatchSnapshot();
  });

  it('should decrement should create decrement action', () => {
    expect(actions.decrement('body')).toMatchSnapshot();
  });

  it('should reset should create reset action', () => {
    expect(actions.reset()).toMatchSnapshot();
  });

  it('should minusSet should create minusSet action', () => {
    expect(actions.minusSet()).toMatchSnapshot();
  });
});
