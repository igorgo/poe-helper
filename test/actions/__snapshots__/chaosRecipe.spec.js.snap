// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions should decrement should create decrement action 1`] = `
Object {
  "payload": Object {
    "itemType": "body",
  },
  "type": "DEC_CHAOS_RECIPE",
}
`;

exports[`actions should increment should create increment action 1`] = `
Object {
  "payload": Object {
    "itemType": "body",
  },
  "type": "INC_CHAOS_RECIPE",
}
`;

exports[`actions should minusSet should create minusSet action 1`] = `
Object {
  "type": "MINUS_SET_CHAOS_RECIPE",
}
`;

exports[`actions should reset should create reset action 1`] = `
Object {
  "type": "RESET_CHAOS_RECIPE",
}
`;
